src/_guides/language/evolution.md

```diff
@@ -127,25 +151,99 @@ void main() {
 }
 ```
 
-## Dart 2.8
+### Dart 2.8
 
 Dart 2.8 didn't add any features to the Dart language, but it did
 contain a number of preparatory [breaking changes][2.8 breaking changes] to ensure great
 nullability-related usability and performance in the upcoming
-null safety feature.
+[null safety][] feature.
 
 It also contained a faster pub tool, and a new [pub outdated][] command.
 
-## Dart 2.9
+### Dart 2.9
 
 Dart 2.9 didn't add any features to the Dart language.
 
-## Dart 2.10
+### Dart 2.10
 
 Dart 2.10 didn't add any features to the Dart language,
 but it added an expanded [`dart` tool][dart-tool] that's 
 analogous to the Flutter SDK's `flutter` tool.
 
+## Language versioning
+
+Before language versioning,
+Dart tools assumed that all Dart code could use all features
+that were supported by the tool's SDK version.
+This assumption causes problems when a package
+that uses a _newer_ language feature
+has [SDK constraints][] that are too low.
+For example, consider a package that uses set literals (2.2)
+but has the SDK constraints `">=2.0.0 <3.0.0"`.
+People using Dart 2.0.0 or 2.1.0 can download the package,
+but their tools can't understand the set literals.
+
+With language versioning,
+a single Dart SDK simultaneously supports
+multiple different versions of the Dart language.
+When you compile some Dart code,
+Dart figures out what version the code is targeting,
+and you're warned if the code uses a feature
+that was introduced after the version in the lower SDK constraint.
+
+On rare occasions, Dart introduces a
+backwards-incompatible feature like [null safety][].
+Previously valid code might not work with null safety,
+so to get the benefits of null safety you'll need to migrate your code.
+And because migrating to null safety can't be completely automated,
+you'll need to be able to migrate one library
+(usually one Dart file) at a time.
```

Comment:
It can't be completely automated, but using the migration tool, the normal workflow is still to migrate an entire package at a time.

The file comments are intended to be a fairly rarely-used feature for most application developers.